blobby/src/lib.rs

```diff
@@ -1,206 +1,248 @@
-//! example "blobby1" means that 1 byte is used for length, and "blobby2",
-//! "blobby4", "blobby8" mean 2, 4 and 8 bytes respectively (this number will
-//! be denoted as `n`).
+//! Storage format represents a sequence of binary blobs. The format uses
+//! git-flavored [variable-length quantity][0] (VLQ) for encoding unsigned
+//! numbers.
 //!
-//! After header goes a sequence of records. Each record starts with prefix of
-//! `n` bytes in which record length (excluding prefix) is stored using little
-//! endian format.
+//! File starts with a number of de-duplicated blobs `d`. It followed by `d`
+//! entries. Each entry starts with a integer `m`, immediately folowed by `m`
+//! bytes representing de-duplicated binary blob.
+//!
+//! Next follows unspecified number of entries representing sequence of stored
+//! blobs. Each entry starts with an unsigned integer `n`. The least significant
+//! bit of this integer is used as a flag. If the flag is equal to 0, then the
+//! number is followed by `n >> 1` bytes, representing a stored binary blob.
+//! Otherwise the entry references a de-duplicated entry number `n >> 1` stored
+//! at the beginning of the storage.
 //!
 //! # Examples
-//! First line represents input binary string and second line shows blobs stored
-//! in it.
-//! ```text
-//! "blobby1\x05hello\x01 \x00\x05world"
-//! "hello", " ", "", "world"
 //! ```
-//! Example for `n=2`, note that length is given in little endian format:
-//! ```text
-//! "blobby2\x03\x00foo\x00\x00\x03\x00bar"
-//! "foo", "", "bar"
+//! let buf = b"\x02\x05hello\x06world!\x01\x02 \x00\x03\x06:::\x03\x01";
+//! let mut v = blobby::BlobIterator::new(buf);
+//! assert_eq!(v.next().unwrap(), b"hello");
+//! assert_eq!(v.next().unwrap(), b" ");
+//! assert_eq!(v.next().unwrap(), b"");
+//! assert_eq!(v.next().unwrap(), b"world!");
+//! assert_eq!(v.next().unwrap(), b":::");
+//! assert_eq!(v.next().unwrap(), b"world!");
+//! assert_eq!(v.next().unwrap(), b"hello");
+//! assert_eq!(v.next(), None);
 //! ```
+//! [0]: https://en.wikipedia.org/wiki/Variable-length_quantity
+#![no_std]
 #![doc(html_logo_url =
     "https://raw.githubusercontent.com/RustCrypto/meta/master/logo_small.png")]
-#![no_std]
 
-use core::{iter::Iterator, convert::TryInto};
+use core::iter::Iterator;
+
+const INDEX_SIZE: usize = 160;
 
 /// Iterator over binary blobs
 pub struct BlobIterator<'a> {
     data: &'a [u8],
-    size: IndexSize,
-}
-
-impl<'a> BlobIterator<'a> {
-    /// Create a new `BlobIterator` for given `data`.
-    pub fn new(data: &'a [u8]) -> Result<Self, &'static str> {
-        if data.len() < 7 { Err("data is too small")? }
-        let (header, data) = data.split_at(7);
-        let size = match header {
-            b"blobby1" => IndexSize::N8,
-            b"blobby2" => IndexSize::N16,
-            b"blobby4" => IndexSize::N32,
-            b"blobby8" => IndexSize::N64,
-            _ => Err("invalid data header")?,
+    idx: [&'a [u8]; INDEX_SIZE],
```

Comment:
I don't quite like this part, but I see only two alternatives:
- Sacrifice no-heap property and store this index on the heap.
- Traverse the index blobs each time, which is obviously much worse performance wise.

@tarcieri 
What do you think?

Reply:
I'm not sure I understand the format or the problem well enough, but I can not that a well-designed format can be decoded in a single-pass without the need for the heap (which would seem to defeat the point of blobby in the first place?)

Reply:
So initially I was using simple offsets, which can be easily processed in a single pass. But offsets usually bigger than 63 bytes, so they will not fit into one byte. This is why I tried to use an index number of a de-duplicated blob instead. Most often encountered blobs are stored near the beginning, so most reference entries fit nicely into one byte. But it causes a problem: we must store mapping "serial number" -> "position of a blob" somewhere. In other words here I traded easiness of parsing for space minimization.

The current code will waste a certain amount of stack space, but I wonder if it will be a big problem, considering the main use-case of this format.

Reply:
I've changed the code to boxed slice. While dependency on `alloc` is not ideal, it looks like it should not be a big obstacle for most embedded targets. Especially considering that AEAD tests right now either way are dependent on `alloc`.